src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,4058 @@
+
+    def get_decoder(self):
+        return self.input_model.get_decoder()
+
+    def get_input_embeddings(self):
+        return self.input_model.get_input_embeddings()
+
+    # @add_start_docstrings_to_model_forward(SEAMLESS_M4T_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
+    # @add_code_sample_docstrings(
+    #    checkpoint=_CHECKPOINT_FOR_DOC,
+    #    output_type=BaseModelOutputWithPastAndCrossAttentions,
+    #    config_class=_CONFIG_FOR_DOC,
+    # )
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        decoder_input_ids: Optional[torch.LongTensor] = None,
+        decoder_attention_mask: Optional[torch.LongTensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        decoder_head_mask: Optional[torch.Tensor] = None,
+        cross_attn_head_mask: Optional[torch.Tensor] = None,
+        encoder_outputs: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        decoder_inputs_embeds: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        **kwargs,
+    ) -> Union[Seq2SeqLMOutput, Tuple[torch.FloatTensor]]:
+        r"""
+        labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
+            Labels for computing the masked language modeling loss. Indices should either be in `[0, ...,
+            config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored
+            (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
+
+        Returns:
+
+        """
+
+        logger.warning(
+            "This calls `self.input_model.forward`. If you want to generate speech, use the `generate` method."
+        )
+
+        return self.input_model.forward(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            decoder_input_ids=decoder_input_ids,
+            decoder_attention_mask=decoder_attention_mask,
+            head_mask=head_mask,
+            decoder_head_mask=decoder_head_mask,
+            cross_attn_head_mask=cross_attn_head_mask,
+            encoder_outputs=encoder_outputs,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            decoder_inputs_embeds=decoder_inputs_embeds,
+            labels=labels,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            **kwargs,
+        )
+
+    @torch.no_grad()
+    def generate(
```

Comment:
This is quite hacky but as far as I can tell it works - WDYT here @joao about generating with two encoder-decoder models like this, one after the other?

Reply:
hey @gante, do you have any insights on this ?

Reply:
(answered on slack)

Reply:
Copying the discussion from Slack for bookkeeping

Questions from @ylacombe:
1. There are 4 tasks and 5 models (1 per tasks and one that encompass every tasks), and the generation config might change from a model to another, how would you define the generation config? Note that ultimately, I'd like to have one repo per model version (independently from the task)   - see this [comment](https://github.com/huggingface/transformers/pull/25693#discussion_r1304177118) as well
2. Models that generates speech (3 out of 5) must do 2 generation calls in a row (similarly to Bark for example). Is is still okay to call the method generate  and are [there](https://github.com/huggingface/transformers/blob/d0310af0440bd6788a94b6f3dedc85b777f7cf5b/src/transformers/models/seamless_m4t/modeling_seamless_m4t.py#L2916-L2942) any particular issues/things to be cautious about? See here for a code example and here for the related [comment](https://github.com/huggingface/transformers/pull/25693#discussion_r1304186733).
3. A following question would be with regards to how to properly test those models with a particular .generate . A bunch of tests of GenerationTesterMixin compare .generate which calls the override generate with other methods that directly calls forward to generate (e.g greedy_search ). What's the best way to deal with those?

Answers from @gante:

1. Generation config is like the model config, in the sense that by default it loads the one with the default name (in from_pretrained ). However, contrarily to the model config, nothing prevents you from changing it after loading the model. You could, for instance, define one generation config per task and, in the custom generation function, ensure the right file is loaded at the right time. An example of what CAN’T be done is to stack the multiple configurations under the same file (it would break the validation code in the class, as we’ve seen in BARK, requiring a custom class).
2. Since the repo holds several generation modes, some form of custom code is always needed.
 I see no problems with having multiple generate calls, it is stateless: (for instance, assisted generation does nested generate calls)
3. My advice for tests would be to
    a) not use the mixin, as it is intended for the vanilla generate function
    b) have some tests with a dummy small model and small inputs, and confirm that the results have the expected shapes (this quickly confirms that the interface is working properly)
    c) have a `@slow` integration test, with a clear input and an expected output — preferably long ones, which are more sensible (this confirms that, when changes on the model/generate/config files are made, the behavior stays intact)
    d) In the slow tests, try to use batch size > 1. (if it was a decoder-only model, I would suggest to test left-padding support as well